pkg/api/subscribeWorker.go

```diff
@@ -0,0 +1,196 @@
+		},
+	)
+	dbChan, err := subscription.Start()
+	if err != nil {
+		return nil, err
+	}
+	worker := &subscribeWorker{
+		ctx:                 ctx,
+		log:                 log,
+		dbSubscription:      dbChan,
+		globalListeners:     make([]listener, 0),
+		originatorListeners: make(map[uint32][]listener),
+		topicListeners:      make(map[string][]listener),
+	}
+
+	go worker.start()
+
+	return worker, nil
+}
+
+func (s *subscribeWorker) start() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case new_batch := <-s.dbSubscription:
+			for _, row := range new_batch {
+				s.dispatch(&row)
+			}
+		}
+	}
+}
+
+func (s *subscribeWorker) dispatch(
+	row *queries.GatewayEnvelope,
+) {
+	bytes := row.OriginatorEnvelope
+	env := &message_api.OriginatorEnvelope{}
+	err := proto.Unmarshal(bytes, env)
+	if err != nil {
+		s.log.Error("Failed to unmarshal envelope", zap.Error(err))
+		return
+	}
+	for _, listener := range s.originatorListeners[uint32(row.OriginatorNodeID)] {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default: // TODO(rich) Close and clean up channel
+		}
+	}
+	for _, listener := range s.topicListeners[hex.EncodeToString(row.Topic)] {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default:
+		}
+	}
+	for _, listener := range s.globalListeners {
+		select {
+		case listener <- []*message_api.OriginatorEnvelope{env}:
+		default:
+		}
+	}
+}
+
+func (s *subscribeWorker) addListeners(
+	requests []*message_api.BatchSubscribeEnvelopesRequest_SubscribeEnvelopesRequest,
+) (<-chan []*message_api.OriginatorEnvelope, error) {
+	subscribeAll := false
+	topics := make(map[string]bool, len(requests))
+	originators := make(map[uint32]bool, len(requests))
+
+	if len(requests) > maxSubscriptionsPerClient {
+		// When a client subscribes to too many originators or topics, we treat it as a request to
+		// subscribe to all instead of throwing an error. We rely on the client's existing
+		// filtering logic rather than forcing clients to respond to an error.
+		subscribeAll = true
+	} else {
+		for _, req := range requests {
+			enum := req.GetQuery().GetFilter()
+			if enum == nil {
+				subscribeAll = true
+			}
+			switch filter := enum.(type) {
+			case *message_api.EnvelopesQuery_Topic:
+				if len(filter.Topic) == 0 || len(filter.Topic) > maxTopicLength {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid topic")
+				}
+				topics[hex.EncodeToString(filter.Topic)] = true
+			case *message_api.EnvelopesQuery_OriginatorNodeId:
+				originators[filter.OriginatorNodeId] = true
+			default:
+				subscribeAll = true
+			}
+		}
+	}
+
+	ch := make(chan []*message_api.OriginatorEnvelope, subscriptionBufferSize)
+
+	if subscribeAll {
+		if len(topics) > 0 || len(originators) > 0 {
+			return nil, fmt.Errorf("cannot filter by topic or originator when subscribing to all")
+		}
+		// TODO(rich) thread safety
+		s.globalListeners = append(s.globalListeners, ch)
```

Comment:
should becoming an globalListener remove said client from all topic/origin listeners?

Reply:
The check on L177 means that the client won't have any topic/origin listeners if they are subscribing globally

Reply:
I was assuming incorrectly that there is only one channel per client and we can merge these all together.